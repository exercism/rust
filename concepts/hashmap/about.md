# `HashMap`

A `HashMap` is a container that can be used to store key-value pairs. In other programming languages, this data structure is sometimes called an associative array or hash table. The map contains a set of keys, each of which is mapped to a particular value. The user can retrieve a stored value by passing in the appropriate key. Users can also insert and delete keys and the associated values.

`HashMap` is a generic collection (like most of the collections in the standard library) and thus supports a broad variety of types as keys, including user-defined structs and tuples. Its values can be of any type.

## Creating a `HashMap`

A `HashMap` is created using the `HashMap::new()` function. The below snippet of code creates a `HashMap` with team names mapped to scores.

```rust
use std::collections::HashMap;

let mut scores = HashMap::new();

scores.insert(String::from("Blue"), 10);
scores.insert(String::from("Yellow"), 50);
```

## Accessing values in `HashMap`

If a value is known to exist within a `HashMap`, then it's appropriate to use the indexing operator (`[]`). To retrieve the score of the blue team, for example, we can use `scores["Blue"]`. However, if the appropriate key-value pair does not exist within the hashmap, this operation will panic.

Other than the indexing operator, there are two other ways of accessing values stored in a `HashMap`. The first is using the `get` member function:

```rust
if let Some(blue_score) = scores.get("Blue") {
    println!("Blue scored: {} \n", blue_score);
}
```

`get` accesses the value stored against the requested key. It returns `None` if the supplied key doesn't exist in the `HashMap`, and `Some(value)` if the key exists. For more about Options in Rust, check the Options concept.

The other way to access values stored in a `HashMap` is to use the `entry` method. The `entry` method (or the entry API) returns a reference to the entry of the key value pair in the `HashMap`. The entry represents the state of this key in the Hash. If the key does not exist, then there is no value in the entry (and it allows one to be inserted).

```rust
let mut vote_counter: HashMap<_, usize> = HashMap::new();
let votes = ["Blue", "Red", "Red", "Blue", "Red", "Blue", "Blue"];
for vote in votes {
    let count = vote_counter.entry(vote).or_default();
    *count += 1;
}

println!("{:#?}", vote_counter);
```

This API makes certain common access patterns very convenient; it has an entire concept (Entry API) dedicated to it.

## Efficiency

`HashMap`s are relatively fast, and have amortized constant-time complexity (`O(1)`) for all operations involving a single key.

## Trait Bounds

A `HashMap` is a generic data structure, meaning it supports arbitrary types to be used as keys and values, with only one constraint: for a type to be used as the key type, it must have an implementation for two traits: `Eq` and `Hash`. There are no trait bounds on the value type.
