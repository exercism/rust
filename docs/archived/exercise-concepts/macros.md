# Concepts Required for the Macros exercise

There are many ways to potentially solve an exercise, but I'll be looking at the [canonical example](https://github.com/exercism/rust/blob/666c366fc0d39d9f6fd8e837facea8afe2153ae2/exercises/macros/example.rs) as my primary reference for what concepts are used.

When referring to the list of concepts, we mean [this list](https://github.com/exercism/v3/blob/master/languages/rust/reference/README.md)

Note that for both concept and practice exercises, the list of prerequisite concepts should be as small as possible. The goal here is not to enumerate all concepts which may be used in all possible approaches to the problem; it is to enumerate the minimal set of knowledge with which a determined student could conceivably solve the problem.

Note that this exercise is very likely to become a concept exercise for the concept "Writing declarative macros".

## Required concepts

### Existing

- structs
- (im)mutability
- `Hashmap<K, V>`

### Should be added

- variable assignment (part of `fundamentals`)
- variable scope / namespacing (part of `fundamentals`)
- methods and associated functions (depends on `structs`)

## Optional Concepts

### Existing

_none_

### Should be added

_none_
