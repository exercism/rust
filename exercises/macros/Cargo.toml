[package]
edition = "2018"
name = "macros"
version = "0.1.0"

[dependencies]

[dev-dependencies]
trybuild = "1.0"  # required to check for invalid syntax
