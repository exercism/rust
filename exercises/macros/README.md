# Macros

Macros are a powerful part of a Rust programmer's toolkit, and [macros by example](https://doc.rust-lang.org/reference/macros-by-example.html) are a relatively simple way to access this power. Let's write one!

## Context

What is a macro? [Wikipedia](https://en.wikipedia.org/wiki/Macro_(computer_science)) describes it thus:

> A macro (short for "macroinstruction", from Greek μακρός 'long') in computer science is a rule or pattern that specifies how a certain input sequence (often a sequence of characters) should be mapped to a replacement output sequence (also often a sequence of characters) according to a defined procedure. The mapping process that instantiates (transforms) a macro use into a specific sequence is known as macro expansion.

Illuminating! But to be more concrete, macros are a special syntax which allows you to generate code at compile time. Macros can be used for compile-time calculation, but more often they're just another way to abstract your code. For example, you've probably already used `println!()` and `vec![]`. These each take an arbitrary number of arguments, so you can't express them as simple functions. On the other hand, they always expand to some amount of absolutely standard Rust code. If you're interested, you can use the [cargo expand](https://github.com/dtolnay/cargo-expand) subcommand to view the results of macro expansion in your code.

For further information about macros in Rust, The Rust Book has a [good chapter](https://doc.rust-lang.org/book/ch19-06-macros.html) on them.

## Problem Statement

You can produce a `Vec` of arbitrary length inline by using the `vec![]` macro. However, Rust doesn't come with a way to produce a [`HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html) inline. Rectify this by writing a `hashmap!()` macro.

For example, a user of your library might write `hashmap!('a' => 3, 'b' => 11, 'z' => 32)`. This should expand to the following code:

```rust
{
   let mut hm = HashMap::new();
   hm.insert('a', 3);
   hm.insert('b', 11);
   hm.insert('z', 32);
   hm
}
```

Note that the [`maplit` crate](https://crates.io/crates/maplit) provides a macro which perfectly solves this exercise. Please implement your own solution instead of using this crate; please make an attempt on your own before viewing its source.

## Rust Installation

Refer to the [exercism help page][help-page] for Rust installation and learning
resources.

## Writing the Code

Execute the tests with:

```bash
$ cargo test
```

All but the first test have been ignored. After you get the first test to
pass, open the tests source file which is located in the `tests` directory
and remove the `#[ignore]` flag from the next test and get the tests to pass
again. Each separate test is a function with `#[test]` flag above it.
Continue, until you pass every test.

If you wish to run all ignored tests without editing the tests source file, use:

```bash
$ cargo test -- --ignored
```

To run a specific test, for example `some_test`, you can use:

```bash
$ cargo test some_test
```

If the specific test is ignored use:

```bash
$ cargo test some_test -- --ignored
```

To learn more about Rust tests refer to the [online test documentation][rust-tests]

Make sure to read the [Modules][modules] chapter if you
haven't already, it will help you with organizing your files.

## Further improvements

After you have solved the exercise, please consider using the additional utilities, described in the [installation guide](https://exercism.io/tracks/rust/installation), to further refine your final solution.

To format your solution, inside the solution directory use

```bash
cargo fmt
```

To see, if your solution contains some common ineffective use cases, inside the solution directory use

```bash
cargo clippy --all-targets
```

## Submitting the solution

Generally you should submit all files in which you implemented your solution (`src/lib.rs` in most cases). If you are using any external crates, please consider submitting the `Cargo.toml` file. This will make the review process faster and clearer.

## Feedback, Issues, Pull Requests

The [exercism/rust](https://github.com/exercism/rust) repository on GitHub is the home for all of the Rust exercises. If you have feedback about an exercise, or want to help implement new exercises, head over there and create an issue. Members of the rust track team are happy to help!

If you want to know more about Exercism, take a look at the [contribution guide](https://github.com/exercism/docs/blob/master/contributing-to-language-tracks/README.md).

[help-page]: https://exercism.io/tracks/rust/learning
[modules]: https://doc.rust-lang.org/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[cargo]: https://doc.rust-lang.org/book/ch14-00-more-about-cargo.html
[rust-tests]: https://doc.rust-lang.org/book/ch11-02-running-tests.html

## Source

Peter Goodspeed-Niklaus

## Submitting Incomplete Solutions
It's possible to submit an incomplete solution so you can see how others have completed the exercise.
