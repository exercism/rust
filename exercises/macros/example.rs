// Ignoring the README's injunction, this is heavily based on the implementation from maplit.
// Original source is at https://github.com/bluss/maplit/blob/master/src/lib.rs#L27-L60

#[macro_export]
macro_rules! hashmap {
    ($($key:expr => $value:expr,)+) => { $crate::hashmap!($($key => $value),+) };
    ($($key:expr => $value:expr),*) => {
        {
            let mut _map = ::std::collections::HashMap::new();
            $(
                _map.insert($key, $value);
            )*
            _map
        }
    };
}
