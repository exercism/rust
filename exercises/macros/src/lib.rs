#[macro_export]
macro_rules! hashmap {
    () => {
        unimplemented!()
    };
}
