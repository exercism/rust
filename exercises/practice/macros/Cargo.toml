[package]
edition = "2021"
name = "macros"
version = "0.1.0"

[dependencies]
