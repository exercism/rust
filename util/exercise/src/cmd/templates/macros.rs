{% macro to_literal(value, use_maplit=false) -%}
    {% if value is object -%}
        {% if use_maplit -%}
            hashmap! {
                {% for k, v in value -%}
                {{ self::to_literal(value=k, use_maplit=use_maplit) }} =>
                {{ self::to_literal(value=v, use_maplit=use_maplit) }},
                {% endfor -%}
            }
        {% else -%}
            {
                let mut hm = ::std::collections::HashMap::new();
                    {% for k, v in value -%}
                    hm.insert(
                        {{ self::to_literal(value=k, use_maplit=use_maplit) }},
                        {{ self::to_literal(value=v, use_maplit=use_maplit) }}
                    );
                    {% endfor -%}
                hm
            }
        {% endif -%}
    {% elif value is iterable -%}
        vec![
            {% for element in value -%}
            {{ self::to_literal(value=element, use_maplit=use_maplit) }},
            {% endfor -%}
        ]
    {% elif value is string -%}
        "{{ value }}"
    {% elif value is number -%}
        {{ value }}
    {% else -%}
        None
    {% endif -%}
{% endmacro -%}

{% macro gen_test_fn(case, first_test_case=false) -%}
    {# Need to set up the variables for the template. #}
    {% set description = case.description -%}
    {% set comments = case.comments -%}
    {% set property = case.property -%}
    {% set input = case.input -%}
    {% set expected = case.expected -%}

    {% include "test_fn.rs" %}
{% endmacro generate_test_fn -%}