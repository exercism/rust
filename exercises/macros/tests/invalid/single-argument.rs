use macros::hashmap;

fn main() {
    // a single argument is invalid
    hashmap!('a');
}
