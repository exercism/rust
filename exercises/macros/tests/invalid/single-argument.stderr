error: unexpected end of macro invocation
 --> $DIR/single-argument.rs:5:17
  |
5 |     hashmap!('a');
  |                 ^ missing tokens in macro arguments
